### ISSUE TITLE
Port the agent(s) to Python

### ISSUE THREAD

**Issue by bracesproul** [langchain-ai/open-canvas#39]
It would be great if both agents were ported to Python and included in this repo.

**Comment by siddicky** [comment]
What would the project structure look like. Split it into frontend and backend (langgraph template) design?

**Comment by bracesproul** [comment]
@siddicky I'm thinking:
```
js-agent/
py-agent/
frontend/
```

With `src/agent/` being moved into `js-agent/`

**Comment by snifhex** [comment]
Hey @bracesproul Is someone working on this? If not, I would like to try this

**Comment by siddicky** [comment]
Which agent would you like to work on first?
I created initial project structure and planned to make a PR over the weekend. Though haven't started actual port yet.

**Comment by snifhex** [comment]
@siddicky I was planning to port reflection first.

**Comment by snifhex** [comment]
@bracesproul Do you think it will be good idea to do something like this 

```
frontend/
agent/js
agent/py
agent/prompts.json
```

have prompts in some universal format which js or py can use like json or something or do we want to have them seperate and  duplicate for pormpt content is okay??

**Comment by bracesproul** [comment]
> @bracesproul Do you think it will be good idea to do something like this 
> 
> 
> 
> ```
> 
> frontend/
> 
> agent/js
> 
> agent/py
> 
> agent/prompts.json
> 
> ```
> 
> 
> 
> have prompts in some universal format which js or py can use like json or something or do we want to have them seperate and  duplicate for pormpt content is okay??

@snifhex yeah i like that structure. Only thing id change is I would put the prompts into individual `.txt` files inside a `prompts/` dir (nested inside `agent/`)

This way we don't need to deal with any sort of quote escaping or new lines etc that a json file would require us to deal with.

**Comment by snifhex** [comment]
@bracesproul  Yeah that sounds good.

So final structure would look something like this:

```
frontend/
agent/js
agent/py
agent/prompts/open-canvas.txt
agent/prompts/reflection.txt
```

**Comment by bracesproul** [comment]
> @bracesproul  Yeah that sounds good.
> 
> 
> 
> So final structure would look something like this:
> 
> 
> 
> ```
> 
> frontend/
> 
> agent/js
> 
> agent/py
> 
> agent/prompts/open-canvas.txt
> 
> agent/prompts/reflection.txt
> 
> ```
> 
> 

Yeah exactly. But note there's like 10 diff prompts inside open canvas so maybe even have the structure be:
```
agent/prompts/open-canvas/...

agent/prompts/reflection/...
```

(Reflection only has a single prompt but will def have more in the future)

**Comment by santosshima** [comment]
py-agent in this case would be based in LangGraph? Would love to have this feature

**Comment by IsraelShok** [comment]
Hi @bracesproul, is anyone working on the Python port?
I can try to do this, but I'm not fluent in JS and TS, so I'll need to know what is backend and what's frontend. 
Do you have such a separation you can share?

Thanks

**Comment by bracesproul** [comment]
hey @IsraelShok the only code you need to port is the code inside the `src/agent/**` directory.

**Comment by IsraelShok** [comment]
Another question @bracesproul, Is there an option to disconnect the canvas from external resources (Supabase, Cloud, Smith) for an internal private data solution?
Thanks!

**Comment by phy101** [comment]
I'm working on it. I already port part of the agents. but it still need a lot of revision and debugging.

**Comment by phy101** [comment]
I was able to port all agents to python. I will open a pull request

**Comment by kaittah** [comment]
What's the status now?
